packages/api/src/clickhouse/index.ts

```diff
@@ -663,78 +684,100 @@ export const getMetricsChart = async ({
       : 'name AS group',
   ];
 
-  switch (dataType) {
-    case 'Gauge':
-      selectClause.push(
-        aggFn === AggFn.Count
-          ? 'COUNT(value) as data'
-          : aggFn === AggFn.Sum
-          ? `SUM(value) as data`
-          : aggFn === AggFn.Avg
-          ? `AVG(value) as data`
-          : aggFn === AggFn.Max
-          ? `MAX(value) as data`
-          : aggFn === AggFn.Min
-          ? `MIN(value) as data`
-          : `quantile(${
-              aggFn === AggFn.P50
-                ? '0.5'
-                : aggFn === AggFn.P90
-                ? '0.90'
-                : aggFn === AggFn.P95
-                ? '0.95'
-                : '0.99'
-            })(value) as data`,
-      );
-      break;
-    case 'Sum':
-      selectClause.push(
-        aggFn === AggFn.Count
-          ? 'COUNT(delta) as data'
-          : aggFn === AggFn.Sum
-          ? `SUM(delta) as data`
-          : aggFn === AggFn.Avg
-          ? `AVG(delta) as data`
-          : aggFn === AggFn.Max
-          ? `MAX(delta) as data`
-          : aggFn === AggFn.Min
-          ? `MIN(delta) as data`
-          : `quantile(${
-              aggFn === AggFn.P50
-                ? '0.5'
-                : aggFn === AggFn.P90
-                ? '0.90'
-                : aggFn === AggFn.P95
-                ? '0.95'
-                : '0.99'
-            })(delta) as data`,
-      );
-      break;
-    default:
-      logger.error(`Unsupported data type: ${dataType}`);
-      break;
+  const isRate = isRateAggFn(aggFn);
+
+  if (dataType === 'Gauge' || dataType === 'Sum') {
+    selectClause.push(
+      aggFn === AggFn.Count
+        ? 'COUNT(value) as data'
+        : aggFn === AggFn.Sum
+        ? `SUM(value) as data`
+        : aggFn === AggFn.Avg
+        ? `AVG(value) as data`
+        : aggFn === AggFn.Max
+        ? `MAX(value) as data`
+        : aggFn === AggFn.Min
+        ? `MIN(value) as data`
+        : aggFn === AggFn.SumRate
+        ? `SUM(rate) as data`
+        : aggFn === AggFn.AvgRate
+        ? `AVG(rate) as data`
+        : aggFn === AggFn.MaxRate
+        ? `MAX(rate) as data`
+        : aggFn === AggFn.MinRate
+        ? `MIN(rate) as data`
+        : `quantile(${
+            aggFn === AggFn.P50 || aggFn === AggFn.P50Rate
+              ? '0.5'
+              : aggFn === AggFn.P90 || aggFn === AggFn.P90Rate
+              ? '0.90'
+              : aggFn === AggFn.P95 || aggFn === AggFn.P95Rate
+              ? '0.95'
+              : '0.99'
+          })(${isRate ? 'rate' : 'value'}) as data`,
+    );
+  } else {
+    logger.error(`Unsupported data type: ${dataType}`);
   }
 
+  const rateMetricSource = SqlString.format(
+    `
+SELECT
+  if(
+    runningDifference(value) < 0
+    OR neighbor(_string_attributes, -1, _string_attributes) != _string_attributes,
```

Comment:
I assume this handles the edge cases. I wonder if this map comparison is strict enough or we need to sort keys first

Reply:
Ah that's a good point, do you think it's something we should do in the query or at ingestion time?

Reply:
It looks like sorting is done on the SDK side today. Prometheus looks to operate under the same assumption so I think we're okay here.